cmake/modules/BuildArrow.cmake

```diff
@@ -87,6 +87,17 @@ function(build_arrow)
     list(APPEND arrow_CMAKE_ARGS -DCMAKE_BUILD_TYPE=Release)
   endif()
 
+  if(CMAKE_SYSTEM_NAME MATCHES "Darwin")
+    # MacOS SDK 14 is based on the LLVM 16.
+    # Since LLVM 16 char_traits for non-char types are deprecated.
+    # https://discourse.llvm.org/t/deprecating-std-string-t-for-non-character-t/66779
+    # This is an issue but we don't want to fix it now for arrow,
+    # which loves its basic_string_view<uint8_t> and even <unsigned long>
+    # Hopefully, by the time LLVM18 is out and has this a hard compile error,
+    # arrow will have dealt with it, and we will have updated to a version with the fix
+    list(APPEND arrow_CMAKE_ARGS -DARROW_CXXFLAGS="-Wno-deprecated-declarations")
```

Comment:
this was fixed in https://github.com/apache/arrow/issues/15281 but our submodule still points to an ancient arrow version. i reraised this subject on the mailing list thread https://lists.ceph.io/hyperkitty/list/dev@ceph.io/thread/KP52HE3J62YHR5LTFYKF4L5WFSXOISK3/

regardless, this doesn't seem like a Darwin-specific issue - maybe it should be conditional on compiler/version instead?

Reply:
Yes, it should, but I didn't feel like getting into the compiler version checks in the cmake. TBH, still not sure how to do it best, but most importantly, this didn't seem worth the trouble. As of now, multiple other places require apple SDK 14+, hence the test is equivalent when on apple.

Reply:
like.. what would be the ultimate test here? checking for a compiler version seems tedious cause we don't know for sure which versions of which compilers will generate the warnings.
So, it should be a custom check, compiling a char_traits for non-char types and testing for failure, and also checking that the arrow is of a version that requires this, and then disabling this warning..

However, following that issue you've found, we could at least add a test for arrow version < 13. Should we?

Reply:
i think it's fine, maybe just add a link to `https://github.com/apache/arrow/issues/15281` in the comment and say it can be removed when we bump the submodule to v14